[TACHYON-868] Remote read should manage block locks by the stream

miasantreble opened the pull request:
https://tachyon.atlassian.net/browse/TACHYON-868

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4878/
Test PASSed.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockInStream.java:
```diff
@@ -46,17 +49,37 @@
     // TODO(calvin): Validate these fields.
     mLocation = new InetSocketAddress(location.getHost(), location.getDataPort());
+    mWorkerClient = mContext.acquireWorkerClient(location.getHost());
+    String blockPath = mWorkerClient.lockBlock(blockId);
```
This should be wrapped in a try block in order to prevent an exception here leaking the `workerClient` resource.

calvinjia commented:
@miasantreble Thanks for the update, could you resolve the conflicts with master and add a unit test for the new behavior (ie. start remote read and try to delete and verify the delete cannot go through with the new locking scheme)?

miasantreble commented:
@calvinjia I assume you mean integration test? The only test I found for RemoteBlockInStream is RemoteBlockInStreamIntegrationTest. I added something like the following:
```java
      RemoteBlockInStream is =
          new RemoteBlockInStream(info.getBlockId(), info.getLength(), info.getLocations().get(0)
              .getWorkerAddress());
      is.read();
      mTfs.delete(f);
      is.close();
      Assert.assertNotNull(mTfs.getInfo(f));
```
but the assert failed and seems like the delete succeeded. I tried the same test with master branch and got the same behavior, do you have any comments?

calvinjia commented:
I think you need to do the `assertNotNull` before the close, since the delete will go through after the lock is released by the delete.

miasantreble commented:
@calvinjia I swapped the close and assert but the test still failed in both my branch and master

yupeng9 commented on clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockInStream.java:
```diff
@@ -46,17 +49,37 @@
+    mWorkerClient = mContext.acquireWorkerClient(location.getHost());
+    String blockPath = mWorkerClient.lockBlock(blockId);
+
+    if (null == blockPath) {
```
NIT: we started to have the convention 
```
blockPath
```

miasantreble replied:
done

yupeng9 commented on servers/src/main/java/tachyon/worker/block/BlockDataManager.java:
```diff
@@ -395,16 +395,15 @@ public String readBlock(long userId, long blockId, long lockId) throws NotFoundE
-   * @param lockId The id of the lock on this block
    * @return the block reader for the block
    * @throws NotFoundException if lockId is not found
    * @throws InvalidStateException if userId or blockId is not the same as that in the LockRecord of
```
Can you update the JavaDoc accordingly?

miasantreble replied:
done

yupeng9 commented on servers/src/main/java/tachyon/worker/block/BlockStore.java:
```diff
@@ -192,14 +192,13 @@ void requestSpace(long userId, long blockId, long additionalBytes) throws NotFou
-   * @param lockId the ID of the lock returned by {@link #lockBlock}
    * @return a {@link BlockReader} instance on this block
    * @throws NotFoundException if lockId is not found
    * @throws InvalidStateException if userId or blockId is not the same as that in the LockRecord of
```
Can you update the JavaDoc accordingly?

miasantreble replied:
done

yupeng9 commented:
Thanks for the patch! I left some comments. Also, could you resolve the merge conflicts?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4958/
Test PASSed.

miasantreble commented:
@yupeng9 yes I understand that. I suspect `TachyonFileSystem#delete` is not guarded by the block locks implemented in `TieredBlockStore` so we have to either fix that, or write a different test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/<h2>Failed Tests: <span class='status-failure'>6</span></h2><h3><a name='org.tachyonproject:tachyon-integration-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport'>org.tachyonproject:tachyon-integration-tests</a>: <span class='status-failure'>6</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/RemoteBlockInStreamIntegrationTest/remoteReadLockTest_0_/'><strong>tachyon.client.RemoteBlockInStreamIntegrationTest.remoteReadLockTest[0]</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/RemoteBlockInStreamIntegrationTest/remoteReadLockTest_1_/'><strong>tachyon.client.RemoteBlockInStreamIntegrationTest.remoteReadLockTest[1]</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/RemoteBlockInStreamIntegrationTest/remoteReadLockTest_2_/'><strong>tachyon.client.RemoteBlockInStreamIntegrationTest.remoteReadLockTest[2]</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/RemoteBlockInStreamIntegrationTest/remoteReadLockTest_3_/'><strong>tachyon.client.RemoteBlockInStreamIntegrationTest.remoteReadLockTest[3]</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/RemoteBlockInStreamIntegrationTest/remoteReadLockTest_4_/'><strong>tachyon.client.RemoteBlockInStreamIntegrationTest.remoteReadLockTest[4]</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4968/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/RemoteBlockInStreamIntegrationTest/remoteReadLockTest_5_/'><strong>tachyon.client.RemoteBlockInStreamIntegrationTest.remoteReadLockTest[5]</strong></a></li></ul>
Test FAILed.

calvinjia commented:
@miasantreble Looking at the unit test again, I think the behavior is correct. The metadata of the file is deleted as soon as the delete goes through. However, the file data should still be available to the open readers. You may want to try `read` -> `delete` -> `sleep` (for delete to propagate to the workers) -> `read` -> `close`. And verify the file cannot be accessed after the initial delete and after the close.

miasantreble commented:
@calvinjia thanks for the idea, I've updated the test accordingly. However I'm not sure if the new lock we introduced in the stream is being test here. Is there anything we could do in the test code to verify the lock is indeed functioning as we expected?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4975/<h2>Build result: FAILURE</span></h2>[...truncated 3122 lines...][JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/pom.xml to org.tachyonproject/tachyon-clients/0.8.0-SNAPSHOT/tachyon-clients-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/pom.xml to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/pom.xml to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.8.0-SNAPSHOT/tachyon-parent-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/dependency-reduced-pom.xml to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jarchannel stopped
Test FAILed.

miasantreble commented:
Jenkins, test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4977/<h2>Build result: FAILURE</span></h2>[...truncated 3122 lines...][JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/pom.xml to org.tachyonproject/tachyon-clients/0.8.0-SNAPSHOT/tachyon-clients-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/pom.xml to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/pom.xml to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.8.0-SNAPSHOT/tachyon-parent-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/dependency-reduced-pom.xml to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jarchannel stopped
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/5025/<h2>Build result: FAILURE</span></h2>[...truncated 3463 lines...][JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/pom.xml to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.8.0-SNAPSHOT/tachyon-parent-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/dependency-reduced-pom.xml to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/target/tachyon-shell-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/target/tachyon-shell-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/target/tachyon-shell-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jarchannel stopped
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/5027/
Test PASSed.

calvinjia commented:
@miasantreble I think the test would test that the lock is held, if the lock is not held the delete would go through and the data would be gone as well, meaning reads from the existing stream would fail.

miasantreble commented:
@calvinjia makes sense, thanks for the explanation!

calvinjia commented on integration-tests/src/test/java/tachyon/client/RemoteBlockInStreamIntegrationTest.java:
```diff
@@ -550,4 +552,32 @@ public void seekAroundLocalBlock() throws IOException, TException {
+  /**
+   * Test remote read stream lock in <code>RemoteBlockInStream</code>.
+   */
+  // TODO: figure out why this test fails
```
Is the TODO still relevant?

miasantreble replied:
done

calvinjia commented on integration-tests/src/test/java/tachyon/client/RemoteBlockInStreamIntegrationTest.java:
```diff
@@ -550,4 +552,32 @@ public void seekAroundLocalBlock() throws IOException, TException {
+              .getWorkerAddress());
+      Assert.assertEquals(0, is.read());
+      mTfs.delete(f);
+      Thread.sleep(1000);
```
You can use `CommonUtils.sleepMs` and just sleep for 3 times the heartbeat interval.

miasantreble replied:
done

calvinjia commented on integration-tests/src/test/java/tachyon/client/RemoteBlockInStreamIntegrationTest.java:
```diff
@@ -550,4 +552,32 @@ public void seekAroundLocalBlock() throws IOException, TException {
+      mTfs.delete(f);
+      Thread.sleep(1000);
+      Assert.assertNull(mTfs.getInfo(f));
+      Assert.assertEquals(1, is.read());
```
you may want to test reading at least more than the read buffer size, otherwise the data may not require a remote read to get.

calvinjia commented:
@miasantreble A general comment about the design, I think we want to pass the lock id through the remote read call and restore the lockId parameter in `getBlockReader`, otherwise there is no guarantee on the worker side that the requester has indeed locked the block and has the rights to read it.

miasantreble commented:
@calvinjia I see, so should `RemoteBlockInStream` provide an API to expose the lock ID for the remote reader to pass around?

calvinjia commented:
@miasantreble I think the lockId can just be passed as a param to getBlockReader?

miasantreble commented:
@calvinjia I'm a bit confused, so in `RemoteBlockInStream` where we current acquire the lock, no lockId is returned, only a blockPath is returned. If we want to have the reader to pass the lockId to `getBlockReader`, how would the reader figure out the lockId?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/5141/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.tachyonproject:tachyon-integration-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/5141/org.tachyonproject$tachyon-integration-tests/testReport'>org.tachyonproject:tachyon-integration-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/5141/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.client/IsolatedTachyonFileSystemIntegrationTest/lockBlockTest3/'><strong>tachyon.client.IsolatedTachyonFileSystemIntegrationTest.lockBlockTest3</strong></a></li></ul>
Test FAILed.

calvinjia commented:
@miasantreble Right, we may need to modify the API to have a `lockBlock` which returns the lockId. Currently the `blockPath` returned is unused anyway except to check for failures.

calvinjia commented:
@miasantreble 
I think this ticket has been addressed by #2271, if so, could you close this PR?

calvinjia commented:
@miasantreble Closing this PR due to #2271, feel free to reopen if you feel like we can make some improvements.
